### PR TITLE
improve handling of metrics

### LINKED ISSUE #20: Server should be stable without metrics configured
Running the server without a monitoring block in the configuration causes a panic:
ex.:
```
    "monitoring": {
        "prometheus_metrics_port": 9273,
        "profiler_port": 4269,
        "profiling_path": "/tmp/trace.out"
    },
```

### PR DESCRIPTION
**Pull request by patrickdemers6**
Improved metrics by:
* Replaced gopkg.in/alexcesaro/statsd.v2 library with github.com/smira/go-statsd to add support for labels.
* New infrastructure allows for adding new metric providers with ease.
* Created a noop metric adapter which does nothing when updating metrics. This is used when no metrics provider is specified.
* Removed metric reporting logic littered throughout the application.

Closes #20

### CODE REVIEW

File: datastore/googlepubsub/publisher.go [hunk 1]
```diff
@@ -81,14 +97,12 @@ func (p *Producer) Produce(entry *telemetry.Record) {
 	})
 	if _, err = result.Get(ctx); err != nil {
 		p.logger.Errorf("pubsub_err err: %v", err)
-		if p.prometheusEnabled {
-			metrics.StatsIncrement(p.statsCollector, "pubsub_publish_total", 1, map[string]string{"record_type": entry.TxType})
-			metrics.StatsIncrement(p.statsCollector, "pubsub_publish_total_bytes", int64(entry.Length()), map[string]string{"record_type": entry.TxType})
-		} else {
-			metrics.StatsIncrement(p.statsCollector, entry.TxType+"_produce", 1, map[string]string{})
-		}
+		metricsRegistry.publishCount.Inc(map[string]string{})
```
Comment (aaronpkahn) [inline datastore/googlepubsub/publisher.go:100]: why do we `metricsRegistry.publishCount.Inc` twice?

File: datastore/kafka/kafka.go [hunk 2]
```diff
@@ -16,13 +18,26 @@ type Producer struct {
...
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
 // NewProducer establishes the kafka connection and define the dispatch method
 func NewProducer(config *kafka.ConfigMap, namespace string, reliableAckWorkers int,
-	ackChan chan (*telemetry.Record), prometheusEnabled bool, statsCollector metrics.MetricCollector, logger *logrus.Logger) (telemetry.Producer, error) {
+	ackChan chan (*telemetry.Record), prometheusEnabled bool, metricsCollector metrics.MetricCollector, logger *logrus.Logger) (telemetry.Producer, error) {
+	metricsOnce.Do(func() { registerMetrics(metricsCollector) })
```
Comment (aaronpkahn) [inline datastore/kafka/kafka.go:40]: should we `metricsOnce.Do(func() {` inside of `registerMetrics`?

File: metrics/metrics.go [hunk 3]
```diff
@@ -1,92 +1,183 @@
...
+		flushDuration := 100 * time.Millisecond
+		if monitoringConfig.Statsd.FlushPeriod > 0 {
+			flushDuration = time.Duration(monitoringConfig.Statsd.FlushPeriod) * time.Millisecond
+		}
+		collector := statsd.NewCollector(monitoringConfig.Statsd.HostPort, monitoringConfig.Statsd.Prefix, logger, flushDuration)
+		return collector
+	}
+
+	logger.Infoln("config_skipping_empty_metrics_provider")
+	return noop.NewCollector()
```
Comment (aaronpkahn) [inline metrics/metrics.go:72]: Seems unnecessary to use the guard version and the fallback version.  Can we delete one of the two?

File: metrics/metrics.go [hunk 4]
```diff
@@ -1,92 +1,183 @@
...
+	gcPausePerSecond adapter.Gauge
+	gcPerSecond      adapter.Gauge
 }
 
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
 // ReportServerUsage collects server usage data and reports to stats lib
```
Comment (aaronpkahn) [inline metrics/metrics.go:91]: `ReportServerUsage collects server usage data and reports metrics`

File: metrics/metrics.go (hunk 3 above)
Comment (patrickdemers6) [reply metrics/metrics.go:72]: Our friendly Go compiler doesn't pickup on the fact that all possible paths are taken if we remove the fallback case.

What do you think of this for maximizing readability?

```go
func NewCollector(monitoringConfig *MonitoringConfig, logger *logrus.Logger) MetricCollector {
	isPrometheus := monitoringConfig != nil && monitoringConfig.PrometheusMetricsPort > 0
	isStatsd := monitoringConfig != nil && monitoringConfig.Statsd != nil

	if isPrometheus {
		return prometheus.NewCollector()
	} 

	if isStatsd {
		flushDuration := 100 * time.Millisecond
		if monitoringConfig.Statsd.FlushPeriod > 0 {
			flushDuration = time.Duration(monitoringConfig.Statsd.FlushPeriod) * time.Millisecond
		}
		return statsd.NewCollector(monitoringConfig.Statsd.HostPort, monitoringConfig.Statsd.Prefix, logger, flushDuration)
	}

	logger.Infoln("config_skipping_empty_metrics_provider")
	return noop.NewCollector()
}
```

File: datastore/googlepubsub/publisher.go [hunk 5]
```diff
@@ -81,14 +97,12 @@ func (p *Producer) Produce(entry *telemetry.Record) {
...
+
+		metricsRegistry.publishCount.Inc(map[string]string{"record_type": entry.TxType})
+		metricsRegistry.publishBytesTotal.Add(int64(entry.Length()), map[string]string{"record_type": entry.TxType})
 	} else {
-		metrics.StatsIncrement(p.statsCollector, "pubsub_err", 1, map[string]string{})
+		metricsRegistry.errorCount.Inc(map[string]string{})
```
Comment (patrickdemers6) [inline datastore/googlepubsub/publisher.go:105]: For error count, is there any reason to not report the "record_type"? @aaronpkahn

File: metrics/metrics.go (hunk 3 above)
Comment (aaronpkahn) [reply metrics/metrics.go:72]: looks good

File: datastore/googlepubsub/publisher.go (hunk 5 above)
Comment (aaronpkahn) [reply datastore/googlepubsub/publisher.go:105]: yeah we should add the record type

**Review by aaronpkahn — APPROVED** [review]

**Pull request merged**